bundles/remote_services/discovery_zeroconf/src/discovery_zeroconf_announcer.c

```diff
@@ -0,0 +1,533 @@
+    free(announcer);
+    return;
+}
+
+static void discoveryZeroconfAnnouncer_eventNotify(discovery_zeroconf_announcer_t *announcer) {
+    eventfd_t val = 1;
+    eventfd_write(announcer->eventFd, val);
+    return;
+}
+
+static bool isLoopBackNetInterface(int ifIndex) {
+    if (ifIndex <= 0) {
+        return false;
+    }
+    bool loopBack = false;
+    int fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (fd >= 0) {
+        struct ifreq ifr;
+        memset(&ifr, 0, sizeof(ifr));
+        if (if_indextoname((unsigned int)ifIndex, ifr.ifr_name) != NULL) {
+            if (ioctl(fd, SIOCGIFFLAGS, &ifr) == 0) {
+                loopBack = !!(ifr.ifr_ifru.ifru_flags & IFF_LOOPBACK);
+            }
+        }
+        close(fd);
+    }
+    return loopBack;
+}
+
+static  celix_status_t discoveryZeroconfAnnouncer_endpointAdded(void *handle, endpoint_description_t *endpoint, char *matchedFilter) {
+    (void)matchedFilter;//unused
+    celix_status_t status = CELIX_SUCCESS;
+    discovery_zeroconf_announcer_t *announcer = (discovery_zeroconf_announcer_t *)handle;
+    assert(announcer != NULL);
+    if (endpointDescription_isInvalid(endpoint)) {
+        celix_logHelper_error(announcer->logHelper, "Announcer: Endpoint is invalid.");
+        return CELIX_ILLEGAL_ARGUMENT;
+    }
+
+    celix_logHelper_info(announcer->logHelper, "Announcer: Add endpoint for %s(%s).", endpoint->serviceName, endpoint->id);
+
+    announce_endpoint_entry_t *entry = calloc(1, sizeof(*entry));
+    assert(entry != NULL);
+    entry->registerRef = NULL;
+    entry->announced = false;
+    entry->uid = celix_utils_stringHash(endpoint->id);
+    entry->ifIndex = (int)celix_properties_getAsLong(endpoint->properties, RSA_DISCOVERY_ZEROCONF_SERVICE_ANNOUNCED_IF_INDEX, DZC_SERVICE_ANNOUNCED_IF_INDEX_DEFAULT);
+    // If it is a loopback interface,we will announce the service on the local only interface.
+    // Because the mDNSResponder will skip the loopback interface,if it found a normal interface.
+    entry->ifIndex = isLoopBackNetInterface(entry->ifIndex) ? kDNSServiceInterfaceIndexLocalOnly : entry->ifIndex;
+    const char *serviceSubType = celix_properties_get(endpoint->properties, DZC_SERVICE_TYPE_KEY, NULL);
+    if (serviceSubType != NULL) {
+        int bytes = snprintf(entry->serviceType, sizeof(entry->serviceType), DZC_SERVICE_PRIMARY_TYPE",%s", serviceSubType);
+        if (bytes >= sizeof(entry->serviceType)) {
+            celix_logHelper_error(announcer->logHelper, "Announcer: Please reduce the length of service type for %s.", serviceSubType);
+            status = CELIX_ILLEGAL_ARGUMENT;
+            goto service_type_err;
+        }
+    } else {
+        CELIX_BUILD_ASSERT(sizeof(entry->serviceType) >= sizeof(DZC_SERVICE_PRIMARY_TYPE));
+        strcpy(entry->serviceType, DZC_SERVICE_PRIMARY_TYPE);
+    }
+    entry->properties = celix_properties_copy(endpoint->properties);
+
+    //Remove properties that mDNS txt record does not need
```

Comment:
If the properties are not needed for discovery, I think they should not be part of the endpoint description.

IMO on which interfaces zeroconf discovery works (RSA_DISCOVERY_ZEROCONF_SERVICE_ANNOUNCED_IF_INDEX) can be a config configuration of discovery_zeroconf.